src/main/java/com/ncc/aif/AIFProgressMonitor.java

```diff
@@ -0,0 +1,108 @@
+package com.ncc.aif;
+
+import ch.qos.logback.classic.Logger;
+import org.topbraid.jenax.progress.NullProgressMonitor;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Progress monitor for AIF validation.
+ * Note that this class is implemented based on the way the TopBraid validator engine sends progress monitor updates.
+ * Writes a tab-delimited file in the following format:
+ * <code>Shape# | Shape Name | Start Time | End Time | Duration (ms)</code>
+ * <code>1 | sh:MinExclusiveConstraintComponent | t1 | t2 | 25ms<code>
+ * <code>2 | sh:NotConstraintComponent | t1 | t2 | 1225ms<code>
+ */
+class AIFProgressMonitor extends NullProgressMonitor {
+    private static final boolean LOGGING = false;
+    private static final Logger logger = (Logger) (org.slf4j.LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME));
+    private static final SimpleDateFormat FORMAT = new SimpleDateFormat("HH:mm:ss.SSS");
+    private static final int LOGGING_THRESHOLD = 1000;
+    private BufferedWriter out;
+    private Map<Integer, String> shapeNameMap = new HashMap<>();
```

Comment:
Then how would `worked()` know which shape just completed?  The shapeNum is keeping track of which shape will complete next since `worked()` doesn't indicate this.

Reply:
I would argue that it doesn't currently know which shape just completed. We're assuming that the calls to `worked` are in the same order as the calls to `subtask`. If we can assume this type of insertion order, we might as well take advantage of it by using a List or Queue. If this isn't helpful, we should catch up on Teams

Reply:
Yeah, at the time this was written, I was still guessing how a multi-threaded user of the `AIFProgressMonitor` class would call worked(int) each time a rule was complete, and I thought there was utility in at least storing the order rules were started.  However, now that we've lobotomized the AIFProgressMonitor (single-threaded, only ever designed to work with TopBraid single-threaded validator), we might as well give up all pretense and just store it as a list as you suggest.